### PR TITLE
Put into cache using root entity name

### PR DESCRIPTION
**Pull request by peterkeatingie**
The cache key for reading is created based on the entity root name but the cache was written with a key based on the found entity (the child in this case) resulting in cache miss.

Using the same root name when putting fixes the issue. I do believe that any extending instances should have unique identifiers as they will have to share an identifier from the base class. This will mean the key won't be ambiguous - if my thinking is correct.

https://github.com/doctrine/orm/issues/7969

### CODE REVIEW

**Review by SenseException — CHANGES_REQUESTED** [review]

File: lib/Doctrine/ORM/Cache/DefaultQueryCache.php [hunk 1]
```diff
@@ -279,9 +279,11 @@ public function put(QueryCacheKey $key, ResultSetMapping $rsm, $result, array $h
 
         $region = $persister->getCacheRegion();
 
+        $cm = $this->em->getClassMetadata($entityName);
+
         foreach ($result as $index => $entity) {
             $identifier                     = $this->uow->getEntityIdentifier($entity);
-            $entityKey                      = new EntityCacheKey($entityName, $identifier);
+            $entityKey  = new EntityCacheKey($cm->getMetadataValue('rootEntityName'), $identifier);
```
[the remark this answers is not in the document]
Comment (peterkeatingie) [reply lib/Doctrine/ORM/Cache/DefaultQueryCache.php]: At first I called the property of the metadata directly and then changed to the method call in situ, hence remaining inside the loop. I see your point though, no need to call each time. Updated now.

[the remark this answers is not in the document]
Comment (peterkeatingie) [reply lib/Doctrine/ORM/Cache/DefaultQueryCache.php]: Fixed the alignment too.

File: tests/Doctrine/Tests/ORM/Functional/Ticket/DDC7969Test.php [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+namespace tests\Doctrine\Tests\ORM\Functional\Ticket;
+
+use Doctrine\ORM\Cache\Region\DefaultMultiGetRegion;
+use Doctrine\Tests\Models\Cache\Attraction;
+use Doctrine\Tests\Models\Cache\Bar;
+use Doctrine\Tests\ORM\Functional\SecondLevelCacheAbstractTest;
+
+class DDC7969Test extends SecondLevelCacheAbstractTest
+{
+    public function testChildEntityRetrievedFromCache()
```
[the remark this answers is not in the document]
Comment (peterkeatingie) [reply tests/Doctrine/Tests/ORM/Functional/Ticket/DDC7969Test.php]: Updated now.

**Comment by peterkeatingie** [comment]
Is there anything I can do to trigger the travis build again without updating the PR? I don't think the failure is related to anything I added.

**Review by SenseException — APPROVED** [review]
I'd like to have more reviewers from the Doctrine team accepting this PR in case I overlooked a detail in the functionality.

**Comment by peterkeatingie** [comment]
> I'd like to have more reviewers from the Doctrine team accepting this PR in case I overlooked a detail in the functionality.

No problem, do I need to add more reviewers?

File: lib/Doctrine/ORM/Cache/DefaultQueryCache.php [hunk 3]
```diff
@@ -279,9 +279,12 @@ public function put(QueryCacheKey $key, ResultSetMapping $rsm, $result, array $h
...
         $region = $persister->getCacheRegion();
 
+        $cm             = $this->em->getClassMetadata($entityName);
+        $rootEntityName = $cm->getMetadataValue('rootEntityName');
+
         foreach ($result as $index => $entity) {
-            $identifier                     = $this->uow->getEntityIdentifier($entity);
-            $entityKey                      = new EntityCacheKey($entityName, $identifier);
+            $identifier = $this->uow->getEntityIdentifier($entity);
+            $entityKey  = new EntityCacheKey($rootEntityName, $identifier);
```
Comment (beberlei) [inline lib/Doctrine/ORM/Cache/DefaultQueryCache.php]:
```suggestion
            $entityKey  = new EntityCacheKey($cm->rootEntityName, $identifier);
```

File: lib/Doctrine/ORM/Cache/DefaultQueryCache.php [hunk 4]
```diff
@@ -279,9 +279,12 @@ public function put(QueryCacheKey $key, ResultSetMapping $rsm, $result, array $h
 
         $region = $persister->getCacheRegion();
 
+        $cm             = $this->em->getClassMetadata($entityName);
+        $rootEntityName = $cm->getMetadataValue('rootEntityName');
```
Comment (beberlei) [inline lib/Doctrine/ORM/Cache/DefaultQueryCache.php]:
```suggestion
```

You should directly use `$cm->rootEntityName` in line 287, `getMetadataValue` has a different use-case.

**Review by beberlei — CHANGES_REQUESTED** [review]

File: lib/Doctrine/ORM/Cache/DefaultQueryCache.php (hunk 4 above)
Comment (peterkeatingie) [reply lib/Doctrine/ORM/Cache/DefaultQueryCache.php]: Hi @beberlei , I originally did that but then the codesniffer job failed due to accessing a property when presented only with an interface - should we suppress or ignore that?

Comment (beberlei) [reply lib/Doctrine/ORM/Cache/DefaultQueryCache.php]: @peterkeatingie yes, you can add `assert($cm instanceof ClassMetadata);` in line 283, this code gets compiled away in production when assertions are off.

File: tests/Doctrine/Tests/ORM/Functional/Ticket/DDC7969Test.php [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+        if ($region instanceof DefaultMultiGetRegion) {
+            $region->getCache()->flushAll();
+        }
+
+        /** @var Bar $bar */
+        $bar = $this->attractions[0];
+
+        $repository = $this->_em->getRepository(Bar::class);
+
+        $this->assertFalse($this->cache->containsEntity(Bar::class, $bar->getId()));
```
Comment (beberlei) [inline tests/Doctrine/Tests/ORM/Functional/Ticket/DDC7969Test.php]: Can you add a check for the same using the root class name as well?

**Review by beberlei — APPROVED** [review]

**Comment by peterkeatingie** [comment]
Thanks @beberlei  @SenseException. Is there anything I can do expedite this PR or just wait now?

**Comment by beberlei** [comment]
@peterkeatingie i am waiting from someone with more SLC experience to ack. One question I have though, you made this PR towards 2.8.x branch, but I would consider this a bugfix, so this would be good for the 2.7 branch. What was your approach here?

**Comment by peterkeatingie** [comment]
> @peterkeatingie i am waiting from someone with more SLC experience to ack. One question I have though, you made this PR towards 2.8.x branch, but I would consider this a bugfix, so this would be good for the 2.7 branch. What was your approach here?

No problem. Actually I might have this wrong - I thought our project had 2.8 using composer show but maybe got that wrong or maybe that's for dev, I can't remember. Anyway checked the lock file now and it's 2.7 alright. Also was following this advice: https://github.com/doctrine/orm#which-branch-should-i-choose.

So if that should be against 2.7 then I guess I need to branch that and make a new PR?!

**Comment by peterkeatingie** [comment]
Closing this in favour of https://github.com/doctrine/orm/pull/8023

**Pull request closed without merging**